autoload/coqtop.py

```diff
@@ -0,0 +1,346 @@
+        if xml.get('val') == 'in_l':
+            return Inl(parse_value(xml[0]))
+        elif xml.get('val') == 'in_r':
+            return Inr(parse_value(xml[0]))
+        else:
+            assert False, 'expected "in_l" or "in_r" in <union>'
+    elif xml.tag == 'option_state':
+        sync, depr, name, value = map(parse_value, xml)
+        return OptionState(sync, depr, name, value)
+    elif xml.tag == 'option_value':
+        return OptionValue(parse_value(xml[0]))
+    elif xml.tag == 'status':
+        path, proofname, allproofs, proofnum = map(parse_value, xml)
+        return Status(path, proofname, allproofs, proofnum)
+    elif xml.tag == 'goals':
+        return Goals(*map(parse_value, xml))
+    elif xml.tag == 'goal':
+        return Goal(*map(parse_value, xml))
+    elif xml.tag == 'evar':
+        return Evar(*map(parse_value, xml))
+    elif xml.tag == 'xml' or xml.tag == 'richpp':
+        return ''.join(xml.itertext())
+
+def parse_error(xml):
+    return ET.fromstring(re.sub(r"<state_id val=\"\d+\" />", '', ET.tostring(xml)))
+
+def build(tag, val=None, children=()):
+    attribs = {'val': val} if val is not None else {}
+    xml = ET.Element(tag, attribs)
+    xml.extend(children)
+    return xml
+
+def encode_call(name, arg):
+    return build('call', name, [encode_value(arg)])
+
+def encode_value(v):
+    if v == ():
+        return build('unit')
+    elif isinstance(v, bool):
+        xml = build('bool', str(v).lower())
+        xml.text = str(v)
+        return xml
+    elif isinstance(v, str):
+        xml = build('string')
+        xml.text = v
+        return xml
+    elif isinstance(v, int):
+        xml = build('int')
+        xml.text = str(v)
+        return xml
+    elif isinstance(v, StateId):
+        return build('state_id', str(v.id))
+    elif isinstance(v, list):
+        return build('list', None, [encode_value(c) for c in v])
+    elif isinstance(v, Option):
+        xml = build('option')
+        if v.val is not None:
+            xml.set('val', 'some')
+            xml.append(encode_value(v.val))
+        else:
+            xml.set('val', 'none')
+        return xml
+    elif isinstance(v, Inl):
+        return build('union', 'in_l', [encode_value(v.val)])
+    elif isinstance(v, Inr):
+        return build('union', 'in_r', [encode_value(v.val)])
+    # NB: `tuple` check must be at the end because it overlaps with () and
+    # namedtuples.
+    elif isinstance(v, tuple):
+        return build('pair', None, [encode_value(c) for c in v])
+    else:
+        assert False, 'unrecognized type in encode_value: %r' % (type(v),)
+
+coqtop = None
+states = []
+state_id = None
+root_state = None
+
+def kill_coqtop():
+    global coqtop
+    if coqtop:
+        try:
+            coqtop.terminate()
+            coqtop.communicate()
+        except OSError:
+            pass
+        coqtop = None
+
+def ignore_sigint():
+    signal.signal(signal.SIGINT, signal.SIG_IGN)
+
+def escape(cmd):
```

Comment:
i think it is nothing to do with vim. 😅  we can not happily parse xml data using `xml.etree.ElementTree` without escaping these special chars.

![image](https://cloud.githubusercontent.com/assets/6234553/19236937/a1f9e846-8f2c-11e6-9f5e-d198da744ab5.png)